### ISSUE TITLE
Kubernetes operator: Proxy pod not started with --accept-routes causing dropped traffic from subnet-router subnet device

### ISSUE THREAD

**Issue by cubinet-code** [tailscale/tailscale#10684]
## Topology:
10.1.1.x <> Tailscale with subnet route 10.1.1.0/24 <> Tailscale Kubernetes Operator <> Tailscale Kubernetes Operator Proxy Pod

## Problem

When deploying with Kubernetes operator, and trying to access from an IP which sits in a remote subnet handled by a remote subnet router no connectivity can be established.

## Troubleshooting

```
❯ curl 100.96.101.35 (Container Talent IP)

... times out
```

The log of the proxy container also shows **Some peers are advertising routes but --accept-routes is false**

```
2023/12/23 17:58:54 Switching ipn state Starting -> Running (WantRunning=true, nm=true)
2023/12/23 17:58:54 derphttp.Client.Connect: connecting to derp-18 (par)
2023/12/23 17:58:54 magicsock: endpoints changed: 31.19.220.235:59292 (stun), 10.1.0.93:46271 (local)
2023/12/23 17:58:54 control: NetInfo: NetInfo{varies=true hairpin=false ipv6=false ipv6os=false udp=true icmpv4=false derp=#18 portmap= link="" firewallmode="nft"}
Some peers are advertising routes but --accept-routes is false
```

## Fixing connectivity:

After connecting to the proxy pd ts-*** and issuing `tailscale up --accept-routes --reset` connectivity starts to work as expected.

```
❯ curl 100.96.101.35
Hostname: whoami-56647559d7-wrgx5
IP: 127.0.0.1
IP: 10.1.0.90
RemoteAddr: 10.1.0.89:55287
GET / HTTP/1.1
Host: 100.96.101.35
User-Agent: curl/8.4.0
Accept: */*
```

## Suggested fix
Operator needs to start tailscale proxy pods with `TAILSCALE_ARGS: ...--accept-routes` or needs to make TAILSCALE_ARGS configurable through the helm chart.

### Steps to reproduce

10.1.1.100 <> Tailscale with subnet route 10.1.1.0/24 <> Kubernetes Operator <> Kubernetes Operator Proxy Pod

Operator Install:
```
helm upgrade \
  --install \
  tailscale-operator \
  tailscale/tailscale-operator \
  --namespace=tailscale \
  --create-namespace \
  --set-string oauth.clientId=kqqVCR3CNTRL \
  --set-string oauth.clientSecret=tskey-client-kqqVCR3CNTRL-***  \
  --set-string apiServerProxyConfig.mode=true \
  --wait
```

Test container:
```
helm upgrade --install whoami --set-string service.annotations."tailscale\.com\/expose"=true \
   cowboysysop/whoami --create-namespace -n whoami --wait
```

### OS

Linux

### OS version

kubernetes-operator

### Tailscale version

v1.56.1

**Comment by irbekrm** [comment]
Hi @cubinet-code ,

Thanks for opening the issue.

> Tailscale Kubernetes Operator Proxy Pod

Could you please describe your use case (i.e are you deploying ingress/egress proxy) and how does it use a subnet router?

**Comment by cubinet-code** [comment]
Thank you for looking into this. The use case is as following:

Acessing the remote cluster service from a device connected via an IPv4 located in a subnet serviced by a tailscale subnet-router serviced subnet:

10.1.1.x (This is where the client is located) > Tailscale linux instance with subnet route 10.1.1.0/24 > (tailnet) > Tailscale Kubernetes Operator > Tailscale Kubernetes Operator Proxy Pod

The return traffic cannot reach the client, as the Tailscale Kubernetes Operator Proxy Pod does not install the subnet route for 10.1.1.x/24.

Does that explain the use case? Otherwise please let me know what information I can help to clarify further.

best regards,

Ollie.

**Comment by irbekrm** [comment]
Thanks for the extra info!

> ❯ curl 100.96.101.35

Maybe I am missing something- do you run this from the `10.1.1.x` client within the exposed subnet? If so, do you have Tailscale installed on the client device or how is it able to discover `100.96.101.35`?

I am not opposed to making it possible to configure proxies to accept subnet routes.
I think we probably do _not_ want to support configuring `TS_EXTRA_ARGS` via the operator as that would introduce a second way how to configure proxy hostnames, subnet routers, exit nodes for which there already is/is planned an operator config mechanism. We could either introduce another `Service`/`Ingress` annotation (`tailscale.com/accept-routes`) or plug this into the generic proxy configuration work that we have in progress- I'd just like to understand this use case fully to see if there is anything else we can do.

**Comment by cubinet-code** [comment]
> Thanks for the extra info!
> 
> > ❯ curl 100.96.101.35
> 
> Maybe I am missing something- do you run this from the `10.1.1.x` client within the exposed subnet? If so, do you have Tailscale installed on the client device or how is it able to discover `100.96.101.35`?

Yes, correct the client sits on `10.1.1.x`. This machine has a static default route to the first Tailgate node ˙Tailscale linux instance with subnet route 10.1.1.0/24˙. 

>  We could either introduce another `Service`/`Ingress` annotation (`tailscale.com/accept-routes`) or plug this into the generic proxy configuration work that we have in progress

This would perfectly address my use case! I just would like ensure that the proxy node will have the remote subnet routes available so the traffic find it path back to the client.

best regards, Ollie.

**Comment by rhyswilliamsza** [comment]
This would be super useful to enable the below routing as a kubernetes service! 

Kubernetes --> Tailscale --> Machine with Route to `192.168.0.0/24`--> Local Device on `192.168.0.0/24`.

**Comment by cubinet-code** [comment]
> We could either introduce another `Service`/`Ingress` annotation (`tailscale.com/accept-routes`) or plug this into the generic proxy configuration work that we have in progress

Dear Irbe, is there any roadmap or rough release time for the work in progress you mentioned?

best regards, Oliver.

**Comment by cubinet-code** [comment]
No worries, I found PR #10724

**Comment by zhouyuf6741** [comment]
hi @cubinet-code, did you figure out to solve this? I am looking into a similar issue.

**Comment by cubinet-code** [comment]
> hi @cubinet-code, did you figure out to solve this? I am looking into a similar issue.

No, I don't think there is a way yet to set --accept-routes on the proxy in any way through the current manifests or CRDs.

#10724 just added other functionality in regards to exit nodes and proxy subnet router.

@irbekrm mentioned that creating a possibility to set "accept-routes" for proxy nodes is on the todo list though.

**Comment by irbekrm** [comment]
Just a note that we are planning to do this. It is a relatively small piece of work, but it is blocked on https://github.com/tailscale/tailscale/issues/10869 as otherwise there is currently no way to reconfigure the `--accept-routes` value once a proxy device has been created. #10869 is high on our priority list, so I'd expect us to be able to get this done soon-ish.

**Comment by XLordalX** [comment]
What's the status on this? @irbekrm

**Comment by irbekrm** [comment]
> What's the status on this? @irbekrm

It won't come out as part of 1.60 release as I was initially hoping, because we prioritized proxy cluster resource customization and that took longer than expected. So hopefully for the next release.
Thanks for your patience.
Generally we are also happy to review folks' PRs, but this is blocked on #10869 which will require some re-designing of the tailscale image.

**Comment by cubinet-code** [comment]
Hello @irbekrm , 

Thank you for your work on this problem and your commits. Do you have an estimate when this potentially could be merged?

**Comment by jfmontanaro** [comment]
Hi, thanks for adding this! I ran into the same problem a couple of weeks ago and was able to make it work by upgrading to the latest version of the kubernetes operator and adding the appropriate `ProxyClass` and annotations.

More recently I was trying to do the same thing with an egress proxy - that is, I would like applications in my Kubernetes cluster to be able to talk to an IP in a routed subnet, so I'm using the `tailscale.com/tailnet-ip` annotation on an `ExternalName` service. 

However, this doesn't seem to be working. I have confirmed that the subnet route is not present in the proxy pod's route table by exec'ing into the pod and running `ip route show table 52`, and the subnet in question doesn't show up.

I could just be doing something wrong but I figured since this feature is so new I might as well check in. Does the `tailscale.com/proxy-class` annotation work on egress proxy pods?

**Comment by irbekrm** [comment]
> Does the tailscale.com/proxy-class annotation work on egress proxy pods

Belated response 🙏🏼 but the issue likely was that `tailscale.com/proxy-class` needs to be a label, not an annotation

**Comment by irbekrm** [comment]
Closing the issue as this was implemented in 1.68.
